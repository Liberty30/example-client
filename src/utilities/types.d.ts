import { ProfileAnnouncement } from "@dsnp/sdk/core/announcements";
import { ActivityContentProfile } from "@dsnp/sdk/core/activityContent";
import {
  ActivityContent,
  ActivityContentNote,
  ActivityContentPerson,
} from "@dsnp/sdk/core/activityContent";

export type HexString = string;
export type EncryptedString = string;
export type URLString = string;
// ### Feed Data Types ###

// ## Profile ##
export interface Profile extends ActivityContentPerson {
  socialAddress: HexString;
}

export declare type HexString = string;
// ## Note ##
export type NoteAttachmentType = "Image" | "Video";
export type NoteAttachment = {
  mediaType: string;
  type: NoteAttachmentType;
  url: URLString;
};

// ## Feed ##
export interface FeedItem {
  fromAddress: HexString;
  content: ActivityContentNote;
  replies?: FeedItem[];
  blockNumber: number;
  hash?: HexString;
  inbox?: boolean;
  timestamp: number;
  topic?: HexString;
  uri?: URLString | undefined;
  rawContent?: string;
  ddid?: HexString;
  inReplyTo?: HexString;
  tags?: string[] | undefined;
}

// ## Graph ##
export interface Graph {
  dsnpUserId: HexString;
  following: HexString[];
  followers: HexString[];
}

// ## Profile ##
export type Profile = ProfileAnnouncement &
  ActivityContentProfile & { handle: string };

// ## FeedItem ##
export type FeedItem = BroadcastAnnouncement & ActivityContentNote;

// ## Reply ##
export type Reply = ReplyAnnouncement & ActivityContentNote;
