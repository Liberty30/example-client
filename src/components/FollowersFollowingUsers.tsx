import React from "react";
import UserAvatar from "./UserAvatar";
import { Button } from "antd";

interface FollowersFollowingUsersProps {
  selectedListTitle: string;
  tempUserList: any[];
}

const FollowersFollowingUsers = ({
  selectedListTitle,
  tempUserList,
}: FollowersFollowingUsersProps): any => {
  const userList = () => {
    switch (selectedListTitle) {
      case "showFollowers":
        return tempUserList.filter((user) => user.followsMe);
      case "showFollowing":
        return tempUserList.filter((user) => user.following);
      default:
        return [];
    }
  };

  const getList = () => {
    return userList().map((user, index) => {
      const buttonText = user.following ? "Unfollow" : "Follow";
      return (
        <div className="FollowersFollowingUsers__user" key={index}>
          <UserAvatar avatarSize="small" profile={user.profile} />
          <div className="FollowersFollowingUsers__name">{user.name}</div>
          <Button className="FollowersFollowingUsers__button" name={buttonText}>
            {buttonText}
          </Button>
        </div>
      );
    });
  };

  return getList();
};
export default FollowersFollowingUsers;
