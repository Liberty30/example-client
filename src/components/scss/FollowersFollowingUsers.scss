.FollowersFollowingUsers__user {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 6px;
}

.FollowersFollowingUsers__name {
  flex: 1;
  padding-left: 16px;
}

.FollowersFollowingUsers__button {
  border-radius: 6px;
}

.FollowersFollowingUsers__buttonFollowIcon {
  transform: rotate(0deg);
  transition: transform 0.3s;
}

.FollowersFollowingUsers__buttonUnfollowIcon {
  transform: rotate(45deg);
}

.FollowersFollowingUsers__button[name="Unfollow"] {
  &:hover,
  &:active,
  &:focus {
    color: $dark-gray;
    border-color: $dark-orange;
    background-color: $light-gray;
  }
}

.FollowersFollowingUsers__button[name="Follow"] {
  background-color: $dark-gray;
  color: $white;

  &:hover,
  &:active,
  &:focus {
    color: $white;
    border-color: $dark-orange;
    background-color: $dark-gray;
  }
}
