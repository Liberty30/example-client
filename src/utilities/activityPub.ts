import {
  NoteActivityPub,
  ActivityPubAttachment,
  PubType,
} from "./activityPubTypes";
import { HexString } from "./types";
import path from "path";

export const noteToActivityPub = (
  actor: HexString,
  note: string,
  uriList: string[]
): NoteActivityPub => {
  const activityPub: NoteActivityPub = {
    actor,
    "@context": "https://www.w3.org/ns/activitystreams",
    type: "Note",
    content: note,
  };
  if (uriList) {
    if (typeof uriList === "string") {
      uriList = [uriList];
    }
    activityPub["attachment"] = uriList.map(
      (item): ActivityPubAttachment => {
        const extension = path.extname(item).replace(".", "");
        let mediaType;
        let pubType: PubType = "Image";
        const supportedVideoDomains = /youtu.be|youtube.com|vimeo.com/;
        if (supportedVideoDomains.test(item)) {
          pubType = "Video";
          mediaType = "text/html";
        } else {
          switch (extension) {
            case "mp4":
              pubType = "Video";
              mediaType = "video/mp4";
              break;
            case "wmv":
              pubType = "Video";
              mediaType = "video/x-ms-wmv";
              break;
            case "mov":
              pubType = "Video";
              mediaType = "video/quicktime";
              break;
            default:
              mediaType = `image/${extension}`;
          }
        }
        return { type: pubType, mediaType: mediaType, url: item };
      }
    );
  }
  return activityPub;
};

export const postReplyToActivityPub = (
  actor: HexString,
  reply: string,
  parent: HexString
): NoteActivityPub => {
  return {
    "@context": "https://www.w3.org/ns/activitystreams",
    actor,
    type: "Note",
    content: reply,
    inReplyTo: parent,
  };
};
