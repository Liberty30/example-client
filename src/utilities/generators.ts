import { HexString } from "./types"

export const randInt = (max: number): number => {
  return Math.floor(Math.random() * Math.floor(max));
};

/**
 * Generate a randomized hex string of a specific length prefixed by `0x`
 * @param length The length of the hex string excluding the prefix
 */
export const generateHexString = (length: number): HexString => {
  return "0x" + [...Array(length)].map(() => randInt(16).toString(16)).join("");
};
