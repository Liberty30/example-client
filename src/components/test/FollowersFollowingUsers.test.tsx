import React from "react";
import FollowersFollowingUsers from "../FollowersFollowingUsers";
import { shallow } from "enzyme";
import { preFabProfiles } from "../../test/testProfiles";
import { ListStatus } from "../../utilities/enums";

const mockTempUserList = [
  preFabProfiles[0],
  preFabProfiles[1],
  preFabProfiles[2],
  preFabProfiles[3],
];

describe("FollowersFollowingUsers", () => {
  it("renders without crashing", () => {
    expect(() => {
      shallow(
        <FollowersFollowingUsers
          listStatus={ListStatus.FOLLOWERS}
          tempUserList={mockTempUserList}
          notFollowing={[preFabProfiles[1]]}
        />
      );
    }).not.toThrow();
  });

  it("displays correct follow button", () => {
    const component = shallow(
      <FollowersFollowingUsers
        listStatus={ListStatus.FOLLOWERS}
        tempUserList={mockTempUserList}
        notFollowing={[preFabProfiles[0]]}
      />
    );
    expect(
      component.find(".FollowersFollowingUsers__button").first().text()
    ).toContain("Follow");
  });

  it("displays correct unfollow button", () => {
    const component = shallow(
      <FollowersFollowingUsers
        listStatus={ListStatus.FOLLOWERS}
        tempUserList={mockTempUserList}
        notFollowing={[preFabProfiles[0]]}
      />
    );
    expect(
      component.find(".FollowersFollowingUsers__button").at(1).text()
    ).toContain("Unfollow");
    expect(
      component.find(".FollowersFollowingUsers__button").at(2).text()
    ).toContain("Unfollow");
    expect(
      component.find(".FollowersFollowingUsers__button").at(3).text()
    ).toContain("Unfollow");
  });
});
