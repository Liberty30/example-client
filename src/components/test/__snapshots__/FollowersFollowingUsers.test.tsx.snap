// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FollowersFollowingUsers displays correct follow or unfollow button 1`] = `
<Fragment>
  <div
    className="FollowersFollowingUsers__user"
    key="0x1DCODE0000"
  >
    <UserAvatar
      avatarSize="small"
      profile={
        Object {
          "@context": "https://www.w3.org/ns/activitystreams",
          "actor": "0x1DCODE0000",
          "discoverable": true,
          "icon": Object {
            "url": "https://image.shutterstock.com/image-vector/monday-time-sparkle-shine-word-600w-731826949.jpg",
          },
          "id": "http://localhost:3003/api/announce/0x1DCODE0000",
          "name": "Monday January",
          "preferredUsername": "0Monday0",
          "socialAddress": "0x1DCODE0000",
          "summary": "",
          "type": "Person",
          "url": "",
          "walletAddress": "0xCODE0000",
        }
      }
    />
    <div
      className="FollowersFollowingUsers__name"
    >
      Monday January
    </div>
    <Button
      className="FollowersFollowingUsers__button"
      name="Follow"
    >
      Follow
      <div
        className="FollowersFollowingUsers__buttonUnfollowIcon"
      >
        ✕
      </div>
    </Button>
  </div>
  <div
    className="FollowersFollowingUsers__user"
    key="0x1DCODE1111"
  >
    <UserAvatar
      avatarSize="small"
      profile={
        Object {
          "@context": "https://www.w3.org/ns/activitystreams",
          "actor": "0x1DCODE1111",
          "discoverable": true,
          "icon": Object {
            "url": "https://image.shutterstock.com/image-vector/traditional-taco-tuesday-neon-light-600w-1193206603.jpg",
          },
          "id": "http://localhost:3003/api/announce/0x1DCODE1111",
          "name": "Tuesday February",
          "preferredUsername": "1Tuesday1",
          "socialAddress": "0x1DCODE1111",
          "summary": "",
          "type": "Person",
          "url": "",
          "walletAddress": "0xCODE1111",
        }
      }
    />
    <div
      className="FollowersFollowingUsers__name"
    >
      Tuesday February
    </div>
    <Button
      className="FollowersFollowingUsers__button"
      name="Unfollow"
    >
      Unfollow
      <div
        className="FollowersFollowingUsers__buttonUnfollowIcon"
      >
        ✕
      </div>
    </Button>
  </div>
  <div
    className="FollowersFollowingUsers__user"
    key="0x1DCODE2222"
  >
    <UserAvatar
      avatarSize="small"
      profile={
        Object {
          "@context": "https://www.w3.org/ns/activitystreams",
          "actor": "0x1DCODE2222",
          "discoverable": true,
          "icon": Object {
            "url": "https://image.shutterstock.com/image-vector/wednesday-pop-art-illustration-vector-600w-219333010.jpg",
          },
          "id": "http://localhost:3003/api/announce/0x1DCODE2222",
          "name": "Wednesday March",
          "preferredUsername": "2Wednesday2",
          "socialAddress": "0x1DCODE2222",
          "summary": "",
          "type": "Person",
          "url": "",
          "walletAddress": "0xCODE2222",
        }
      }
    />
    <div
      className="FollowersFollowingUsers__name"
    >
      Wednesday March
    </div>
    <Button
      className="FollowersFollowingUsers__button"
      name="Follow"
    >
      Follow
      <div
        className="FollowersFollowingUsers__buttonUnfollowIcon"
      >
        ✕
      </div>
    </Button>
  </div>
  <div
    className="FollowersFollowingUsers__user"
    key="0x1DCODE3333"
  >
    <UserAvatar
      avatarSize="small"
      profile={
        Object {
          "@context": "https://www.w3.org/ns/activitystreams",
          "actor": "0x1DCODE3333",
          "discoverable": true,
          "icon": Object {
            "url": "https://image.shutterstock.com/image-vector/throwback-thursday-brush-lettering-vector-600w-467925458.jpg",
          },
          "id": "http://localhost:3003/api/announce/0x1DCODE3333",
          "name": "Thursday April",
          "preferredUsername": "3Thursday3",
          "socialAddress": "0x1DCODE3333",
          "summary": "",
          "type": "Person",
          "url": "",
          "walletAddress": "0xCODE3333",
        }
      }
    />
    <div
      className="FollowersFollowingUsers__name"
    >
      Thursday April
    </div>
    <Button
      className="FollowersFollowingUsers__button"
      name="Follow"
    >
      Follow
      <div
        className="FollowersFollowingUsers__buttonUnfollowIcon"
      >
        ✕
      </div>
    </Button>
  </div>
</Fragment>
`;
